sequence_processing_pipeline/Pipeline.py

```diff
@@ -166,6 +236,133 @@ def __init__(self, configuration_file_path, run_id, sample_sheet_path,
+                # open each profile and perform minimum validation on its
+                # contents.
+                contents = json_load(f)
+
+                # all files must contain a root element 'profile'. This helps
+                # to identify it as a profile, rather than another type of
+                # JSON file.
+                if 'profile' not in contents:
+                    raise ValueError("'profile' is not an attribute in "
+                                     f"'{profile_path}'")
+
+                # the 'profile' attribute must have a dictionary as its value.
+                # all profiles must contain 'instrument_type' and 'assay_type',
+                # unless instrument_type == 'default', in which case the
+                # profile defines the defaults across all instrument-types and
+                # assay-types.
+                if 'instrument_type' not in contents['profile']:
+                    raise ValueError("'instrument_type' is not an attribute "
+                                     f"in '{profile_path}'.profile")
+
+                if 'assay_type' not in contents['profile']:
+                    if contents['profile']['instrument_type'] != 'default':
+                        raise ValueError("'assay_type' is not an attribute "
+                                         f"in '{profile_path}'.profile")
+
+                profiles.append(contents)
+
+        # The default profile provides 'fall-through' configuration settings
+        # for all items. This allows the user to not have to redefine settings
+        # for all items for all instrument and assay combinations.
+
+        # the final profile is created by taking the default profile and using
+        # it as a base. If a profile matching the run-directory's instrument
+        # and assay types is found, settings from that profile will overwrite
+        # the base-profile settings as appropriate.
+        base_profile = None
+        selected_profile = None
+
+        # iterate through all the profiles, searching for a default
+        # profile and the first profile w/matching instrument and assay types.
+        # if a matching profile isn't found, that's okay, but if a default
+        # profile isn't found, then raise an Error.
+
+        for profile in profiles:
+            p_i_type = profile['profile']['instrument_type']
+            if p_i_type == 'default':
+                base_profile = profile
+            else:
+                p_a_type = profile['profile']['assay_type']
+
+                # if both items have been found, it's safe to break early.
+                if base_profile is not None and selected_profile is not None:
+                    break
+
+                if p_i_type == instr_type and p_a_type == assay_type:
+                    selected_profile = profile
+
+        if base_profile is None:
+            raise ValueError("a 'default' profile was not found")
+
+        if selected_profile:
```

Comment:
should there be an else here?

Reply:
no. The 'default' or base-profile must always be present to be the fall-thru configuration, but if no profile was found that matches the instrument/assay type, that's okay.